runtime/commonMain/src/kotlinx/serialization/Decoding.kt

```diff
@@ -479,39 +486,84 @@ public interface CompositeDecoder {
      */
     public fun decodeStringElement(descriptor: SerialDescriptor, index: Int): String
 
+    @Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE", "DeprecatedCallableAddReplaceWith")
+    @kotlin.internal.LowPriorityInOverloadResolution
+    @Deprecated(decodeMethodDeprecated, level = DeprecationLevel.ERROR)
+    public fun <T : Any?> decodeSerializableElement(
+        descriptor: SerialDescriptor,
+        i: Int, // renamed from index to be called even with LowPriority
+        deserializer: DeserializationStrategy<T>
+    ): T = decodeSerializableElement(descriptor, i, deserializer, null)
+
     /**
-     * Delegates decoding value of the type [T] to the given [deserializer].
+     * Decodes value of the type [T] with the given [deserializer].
+     *
+     * Particular implementations of [CompositeDecoder] may use their format-specific deserializers
+     * for particular data types, e.g. handle [ByteArray] specifically if format is binary.
+     *
+     * If value at given [index] was already decoded with previous [decodeSerializableElement] call with the same index,
+     * [oldValue] would contain a previously decoded value.
+     * Implementation may ignore it and return new value, efficiently overwriting decoded value,
+     * or process it and return merged value if format supports such an operation.
      */
+    @Suppress("DEPRECATION_ERROR")
     public fun <T : Any?> decodeSerializableElement(
         descriptor: SerialDescriptor,
         index: Int,
-        deserializer: DeserializationStrategy<T>
-    ): T
+        deserializer: DeserializationStrategy<T>,
+        oldValue: T? = null
+    ): T = decodeSerializableElement(descriptor, i = index, deserializer = deserializer)
 
     /**
-     * Delegates decoding nullable value of the type [T] to the given [deserializer].
+     * Decodes nullable value of the type [T] with the given [deserializer].
+     *
+     * If value at given [index] was already decoded with previous [decodeNullableSerializableElement] call with the same index,
+     * [oldValue] would contain a previously decoded value.
+     * Implementation may ignore it and return new value, efficiently overwriting decoded value,
+     * or process it and return merged value if format supports such an operation.
      */
+    @Suppress("DEPRECATION_ERROR")
     public fun <T : Any> decodeNullableSerializableElement(
         descriptor: SerialDescriptor,
         index: Int,
+        deserializer: DeserializationStrategy<T?>,
+        oldValue: T? = null
```

Comment:
I think this is important to rename this parameter to `previousValue` and use `previous` terminology everywhere. "previous" is a technical term that clearly reflects the purpose of this parameter